src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+            Contract.ThrowIfNull(parameterList);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in methodSymbol.Parameters)
+            {
+                // Want to skip optional parameters, params parameters, and CancellationToken since they should be at
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private async Task<Solution> RewriteSolutionAsync(
+            Document originalDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, List<SyntaxNode>> callSites, bool trampoline,
+            bool overload, CancellationToken cancellationToken)
+        {
+            var modifiedSolution = originalDocument.Project.Solution;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    if (trampoline || overload)
+                    {
+                        var newRoot = await ModifyDocumentInvocationsTrampolineOverloadAndIntroduceParameterAsync(compilation,
+                            document, originalDocument, invocations, mappingDictionary, methodSymbol, containingMethod,
+                            allOccurrences, parameterName, expression, trampoline, overload, cancellationToken).ConfigureAwait(false);
+                        modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, newRoot);
+                    }
+                    else
+                    {
+                        var newRoot = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                            originalDocument, document, mappingDictionary, methodSymbol, containingMethod,
+                            expression, allOccurrences, parameterName, invocations,
+                            cancellationToken).ConfigureAwait(false);
```

Comment:
not for this PR, but contact me offline.  we should discuss how future PRs should use the 'replace method with object' refactoring to avoid these enormous calls with tons of state passed along.